gtk4: add sample/examples/application1

ToshioCP opened the pull request:
The programs corresponds to the ones in the GTK repository.

- `exampleapp.rb` is almost same as the one under `gtk3` direcotry in the Ruby-GNOME repository.
- It doesn't use `utils.rb` because of the installation. It uses `require "gtk4"` instead.
- `README.md` explains the installation and the desktop file. Its original file is `README` in the application1 directory in the original GTK repository. But some explanation is added for Ruby version.

kou commented on gtk4/sample/examples/application1/README:
```diff
@@ -0,0 +1,25 @@
+Step 1: A trivial application
```
How about renaming this file to `README.md` and use Markdown?

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+#!/bin/sh
+exec ruby -x "$0" "$@"
+#!ruby
+
```
How about simplifying this shebang?

```suggestion
#!/usr/bin/env ruby
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+#!ruby
+
+#
+# Copyright (C) 2022  Ruby-GNOME Project Team
```
```suggestion
# Copyright (C) 2023  Ruby-GNOME Project Team
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+require "gtk4"
+
+class ExampleAppWindow < Gtk::ApplicationWindow
+
+  def open(file)
+    
+  end
```
```suggestion
  def open(file)
  end
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+# example from https://gitlab.gnome.org/GNOME/gtk/blob/main/examples/window-default.c
```
Is this correct?

```suggestion
# Example from:
# * https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleapp.c
# * https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleappwin.c
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+      window.present
+    end
+    signal_connect "open" do |application, files, hin|
+      windows = application.windows
+      win = nil
+      unless windows.empty?
+        win = windows.first
+      else
+        win = ExampleAppWindow.new(application)
+      end
+
+      files.each { |file| win.open(file) }
+        
+      win.present
```
How about simplifying this?

```suggestion
      window = application.windows[0] || ExampleAppWindow.new(application)
      files.each do |file|
        window.open(file)
      end
      window.present
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+end
+
+app = ExampleApp.new
+
+puts app.run
```
Do we need to `puts` here?

```suggestion
app = ExampleApp.new
app.run
```

kou commented on gtk4/sample/examples/application1/README:
```diff
@@ -0,0 +1,25 @@
+  > ~/.local/share/applications/org.gtk.exampleapp.desktop
+
+mkdir -p ~/.local/share/icons/hicolor/48x48/apps
+cp exampleapp.png ~/.local/share/icons/hicolor/48x48/apps
```
Could you also mention the source of `exampleapp.png` ( https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleapp.png ) and its license in this file?

kou reviewed (commented):
Thanks!

ToshioCP commented:
Thank your for the suggestion.

I've corrected mistakes you had pointed and added a bit to the README.md.
I've checked all the files but it might have some mistakes because of  my carelessness.

Toshio Sekiya

kou commented on gtk4/sample/examples/application1/README.md:
```diff
@@ -0,0 +1,29 @@
+# Step 1: A trivial application
+
+The following is only useful in the desktop system based on [Desktop Entry Specification](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html) such as Gnome.
```
```suggestion
The following is only useful in the desktop system based on [Desktop Entry Specification](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html) such as GNOME.
```

kou commented on gtk4/sample/examples/application1/README.md:
````diff
@@ -0,0 +1,29 @@
+
+To make gnome-shell use the desktop file and icon for this example while running it uninstalled, do the following:
+
+```
````
```suggestion
```console
```

kou commented on gtk4/sample/examples/application1/README.md:
````diff
@@ -0,0 +1,29 @@
+```
+
+The file exapmleapp.png is from the GTK GitLab repository.
+The original file is [here](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1) and the license is LGPL.
````
```suggestion
The original file is [here](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleapp.png) and the license is LGPL 2.1 or later.
```

kou commented on gtk4/sample/examples/application1/README.md:
```diff
@@ -0,0 +1,29 @@
+
+1. You need to install your exampleapp.rb in advance.
+
+        $ cp exampleapp.rb $HOME/.local/bin/exampleapp.rb
+        $ chmod 755 $HOME/.local/bin/exampleapp.rb
```
```suggestion
   ```console
   $ cp exampleapp.rb ~/.local/bin/exampleapp.rb
   $ chmod +x ~/.local/bin/exampleapp.rb
   ```
```

kou commented on gtk4/sample/examples/application1/README.md:
```diff
@@ -0,0 +1,29 @@
+        $ chmod 755 $HOME/.local/bin/exampleapp.rb
+
+2. If you use rbenv, your ruby executable may not be found.
+You can solve this by creating a symbolic link in $HOME/.local/bin to your ruby executable.
+For example,
```
```suggestion
   You can solve this by creating a symbolic link in `~/.local/bin` to your ruby executable.
   For example,
```

kou commented on gtk4/sample/examples/application1/README.md:
```diff
@@ -0,0 +1,29 @@
+You can solve this by creating a symbolic link in $HOME/.local/bin to your ruby executable.
+For example,
+
+        $ cd $HOME/.local/bin
+        $ ln -s $HOME/.rbenv/versions/3.1.2/bin/ruby ruby
```
```suggestion
   ```console
   $ cd ~/.local/bin
   $ ln -s $(rbenv which ruby) ruby
   ```
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,49 @@
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+# Example from:
```
```suggestion
#
# Example from:
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,49 @@
+# Example from:
+# * https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleapp.c
+# * https://gitlab.gnome.org/GNOME/gtk/-/blob/main/examples/application1/exampleappwin.c#
+# License: LGPL2.1+
```
How about using SPDX identifier?
https://spdx.org/licenses/

```suggestion
# License: LGPL-2.1-or-later
```

kou commented on gtk4/sample/examples/application1/exampleapp.rb:
```diff
@@ -0,0 +1,49 @@
+end
+
+app = ExampleApp.new
+app.run
```
Could you add a new line here?

ToshioCP commented:
Two files README.md and exampleapp.rb have been modified along your suggestion.

Toshio Sekiya

kou commented on gtk4/sample/examples/application1/README.md:
````diff
@@ -0,0 +1,33 @@
+
+1. You need to install your exampleapp.rb in advance.
+
+```console
+$ cp exampleapp.rb $HOME/.local/bin/exampleapp.rb
+$ chmod 755 $HOME/.local/bin/exampleapp.rb
+```
````
```suggestion
   ```console
   $ cp exampleapp.rb ~/.local/bin/exampleapp.rb
   $ chmod +c ~/.local/bin/exampleapp.rb
   ```
```

kou commented on gtk4/sample/examples/application1/README.md:
````diff
@@ -0,0 +1,33 @@
+   You can solve this by creating a symbolic link in `~/.local/bin` to your ruby executable.
+   For example,
+
+```console
+$ cd $HOME/.local/bin
+$ ln -s $HOME/.rbenv/versions/3.1.2/bin/ruby ruby
+```
````
```suggestion
   ```console
   $ ln -s $(rbenv which ruby) ~/.local/bin/ruby
   ```
```

kou commented:
Thanks!
I've merged with some minor style adjustments.
